File: Day-9: Recursion/1) Subset Sums.cpp
class Solution {
  public:
    vector<int> subsetSums(vector<int>& arr) {
        // code here
        int n = arr.size();
        vector<int> ans;
        helper(0, 0, arr, ans);
        return ans;
    }
    void helper(int id, int sum, vector<int> &nums, vector<int> &ans)
    {
        int n = nums.size();
        if(id == n)
        {
            ans.push_back(sum);
            return;
        }
        helper(id+1, sum + nums[id], nums, ans);
        helper(id+1, sum, nums, ans);
    }
};

File: Day-9: Recursion/1) Subset Sums.java
// User function Template for Java//User function Template for Java
class Solution {
    public ArrayList<Integer> subsetSums(int[] arr) {
        // code here
        List<Integer> ans = new ArrayList<>();
        helper(0, 0, arr, ans);
        return ans;
    }
    public void helper(int index, int sum, int[] nums, List<Integer> ans)
    {
        if(index == nums.length)
        {
            ans.add(sum);
            return;
        }
        helper(index + 1, sum + nums[index], nums, ans);
        helper(index + 1, sum, nums, ans);
    }
}
